packages/lit-dev-content/site/articles/article/redux-reactive-controllers.md

```diff
@@ -0,0 +1,633 @@
+
+```ts
+import type { Store, Action as RAction } from '@reduxjs/toolkit';
+...
+
+export type Selector<State, Result> = (state: State) => Result;
+
+export class SelectorController<
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  private _host: ReactiveControllerHost;
+  private _store: Store<State, Action>;
+  private _selector: Selector<State, Result>;
+
+  constructor(
+    host: ReactiveControllerHost,
+    store: Store<State, Action>,
+    selector: Selector<State, Result>,
+  ) {
+    this._host = host;
+    host.addController(this);
+
+    this._store = store;
+    this._selector = selector;
+  }
+}
+```
+
+Now let’s initialize the initially selected value so that the user can access the state’s initial value with `selectorController.value` using the user's selector:
+
+```ts
+...
+export class SelectorController<
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  ...
+  value: Result;
+
+  constructor(
+    host: ReactiveControllerHost,
+    store: Store<State, Action>,
+    selector: Selector<State, Result>,
+  ) {
+    ...
+    this.value = selector(store.getState());
+  }
+}
+```
+
+### Updating the Component on Store Update
+
+We now have a controller that initializes to the initial state of the store, next let’s update `this.value` when the state updates and then tell the host element to re-render when we have detected a change in the selected value.
+
+Redux stores have a [`Store.subscribe(listener)`](https://redux.js.org/api/store#subscribelistener) method which will call a given callback whenever the state of the store updates. Let's hook into this, update `this.value`, and tell the host to update when the component is connected to the DOM:
+
+```ts
+...
+export class SelectorController<
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  ...
+
+  hostConnected() {
+    this.store.subscribe(() => {
+      const selected = this._selector(this._store.getState());
+      if (selected !== this.value) {
+        this.value = selected;
+        this._host.requestUpdate();
+      }
+    });
+  }
+}
+```
+
+Great! Now the controller will update its value and tell the host element to update whenever the state changes. Additionally, by first comparing the previous state to the current state, we can avoid re-rendering components that don't need to be re-rendered which can improve performance. Nice!
+
+In conclusion, we need to improve our component so that it does not re-render when the component is disconnected from the page and the store’s state changes. Redux’s `Store.subscribe()` method returns an `unsubscribe()` function. Let’s keep track of this and unsubscribe from the store’s changes when the component disconnects.
```

Comment:
Nit: Starting with `In conclusion,` makes me feel like we're about to wrap up the article.